core/src/main/scala/higherkindness/mu/rpc/srcgen/GeneratorApplication.scala

```diff
@@ -37,16 +41,36 @@ class GeneratorApplication[T <: Generator](generators: T*) {
       inputFiles: Set[File],
       outputDir: File
   ): Seq[File] =
-    if (idlTypes.contains(idlType))
-      generatorsByType(idlType).generateFrom(inputFiles, serializationType).map {
-        case (inputFile, outputFilePath, output) =>
-          val outputFile = new File(outputDir, outputFilePath)
-          logger.info(s"$inputFile -> $outputFile")
-          Option(outputFile.getParentFile).foreach(_.mkdirs())
-          outputFile.write(output)
-          outputFile
+    if (idlTypes.contains(idlType)) {
+      val result: ValidatedNel[(File, NonEmptyList[Error]), List[File]] =
+        generatorsByType(idlType)
+          .generateFrom(inputFiles, serializationType)
+          .traverse {
+            case (inputFile, outputFilePath, output) =>
+              output match {
+                case Invalid(readErrors) =>
+                  (inputFile, readErrors).invalidNel
+                case Valid(content) =>
+                  val outputFile = new File(outputDir, outputFilePath)
+                  logger.info(s"$inputFile -> $outputFile")
+                  Option(outputFile.getParentFile).foreach(_.mkdirs())
+                  outputFile.write(content)
+                  outputFile.validNel
+              }
+          }
+      result match {
+        case Invalid(listOfFilesAndReadErrors) =>
+          val formattedErrorMessage = listOfFilesAndReadErrors.map {
+            case (inputFile, errors) =>
+              s"$inputFile has the following errors: ${errors.toList.mkString(", ")}"
+          }
+          throw new RuntimeException(
+            s"One or more IDL files are invalid. Error details:\n $formattedErrorMessage"
+          )
+        case Valid(outputFiles) =>
+          outputFiles
```

Comment:
@dmarticus we're creating some files even in the case we're getting some errors. What do you think about performing the effects just when all files has been validated? Something like:

```scala
val result: ValidatedNel[(File, NonEmptyList[Error]), List[(File, List[String])]] =
//...
  case Valid(content) =>
    val outputFile = new File(outputDir, outputFilePath)
    logger.info(s"$inputFile -> $outputFile")
    (outputFile, content).validNel
//...
result match {
//...
  case Valid(outputFiles) =>
    outputFiles.map {
      case (outputFile, content) =>
        Option(outputFile.getParentFile).foreach(_.mkdirs())
        outputFile.write(content)
        outputFile
    }
```

Reply:
oh good call!  That way we're not wasteful with I/O.

Reply:
made the change :)